proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,618 @@
+
+When ``-XOverloadedStrings`` is enabled, string interpolation can be used for any ``Buildable`` type. Otherwise, it will only ever build Strings.
+
+Interpolation is also supported with ``-XMultilineStrings``, as described in "Proposed Change Specification".
+
+``Buildable`` *could* subsume ``IsString``, so that OverloadedStrings instead desugars string literals to ``fromBuilder (toBuilder s)``. However, it's a better user experience to allow toggling these independently. It's reasonable for someone to want to write interpolated strings, but monomorphize it to String and not require specifying the result type.
+
+Costs and Drawbacks
+-------------------
+
+Development should be low-effort, maintenance should be low-effort. Learnability for novice users will go up, since novice users probably expect string interpolation to be available, and might be frustrated at the lack of support currently.
+
+The major drawback of this approach is the typeclass instances problem:
+
+#. A new interpolator type (e.g. ``SqlQuery``) needs to define ``Builder`` and ``Interpolate`` for all known interpolatable types
+#. A new interpolatable type (e.g. ``BigDecimal``) needs to define ``Interpolate`` for all known interpolator types
+
+This is worse than ``IsString`` or ``Show`` due to the multi-param ``Interpolate`` type class. This makes ``Interpolate`` much more susceptible to orphan instances.
+
+One minor drawback is the whitespace sensitivity of ``s"``, as discussed in "Effect and Interactions".
+
+Alternatives
+------------
+
+* Status quo (discussed in the "Motivation" section)
+
+* Allow ``$foo`` in addition to ``${foo}``
+
+  * This would complicate the syntax, and would also require interpolated string to escape bare ``$``.
+
+* Different delimiter
+
+  * Could use ``f"`` like Python, with ``f`` for format. ``s`` for "String" seems a bit ad-hoc, but it does "look better" for some reason. ``s`` is also a bit better if the user forgets to enable ``-XStringInterpolation`` because ``f`` is a not-uncommon name for functions and ``f"asdf"``, being parsed as ``f "asdf"``, would work more often than ``s"asdf"`` would.
+  * Could reuse QuasiQuote syntax, e.g. ``[s|`` or ``[fmt|``, except it would be special and NOT use Template Haskell.
+
+* No delimiter, always interpolate
+
+  * Would require any use of ``${...}`` to be escaped.
+  * No other language does this; even Bash has single quoted strings to avoid escaping
+
+* Different interpolation delimiter, e.g. ``#{foo}``
+
+  * Most languages use ``$``, and I see no reason to deviate
+
+* Only allow interpolating string-like values
+
+  * This is what ``neat-interpolation`` does
+  * This would add a ton of noise to string interpolation, so no one would use the feature
+  * This wouldn't support injection-free SqlQuery, as you need to know which SqlValue to use
+  * This wouldn't support escaping HTML by default, while allowing explicitly marking certain strings as safe raw HTML
```

Comment:
ah I misunderstood your argument. You're saying that instead of
```hs
s"a ${x}" == fromBuilder (toBuilder "a " <> interpolate x)
```
you want just
```hs
s"a ${x}" == fromBuilder (toBuilder "a " <> x)
```
When you initially said "string type being interpolated", I thought you were implying that everything would interpolate as String.

Yes, I agree that'd still be safe, and is a perfectly reasonable alternative. It would still be better than what we have today. However, I do think it'd be nicer to have this out of the box, and the user could always go the `stringToHTML` route if they wanted to avoid the implicit conversions. So I'd slightly prefer providing `Interpolate`, but if the committee would reject it otherwise, I'd be fine removing it. What's the best way to encode that in the proposal?

Reply:
I mostly just want it clearly described as an alternative, since I think it's an important one to at least _think_ about! I think you get what I was going at now, so I'm happy for you to just describe it :)

Reply:
To be clear, what I'm proposing would be
```
s"a ${x}" == fromBuilder (toBuilder "a " <> toBuilder x)
```
i.e. you interpolate in values of the final type, not the builders!